pylint/checkers/base/basic_checker.py

```diff
@@ -317,6 +317,20 @@ def _check_using_constant_test(
         emit = isinstance(test, (nodes.Const,) + structs + const_nodes)
         if not isinstance(test, except_nodes):
             inferred = utils.safe_infer(test)
+        # Emit if calling a function that only returns GeneratorExp (always tests True)
+        elif isinstance(test, nodes.Call):
+            inferred_call = utils.safe_infer(test.func)
+            if isinstance(inferred_call, nodes.FunctionDef):
+                all_returns_were_generator = None
+                for return_node in inferred_call._get_return_nodes_skip_functions():
+                    if not isinstance(return_node.value, nodes.GeneratorExp):
+                        all_returns_were_generator = False
+                        break
+                    all_returns_were_generator = True
```

Comment:
```suggestion
                if any(
                    not isinstance(return_node.value, nodes.GeneratorExp):
                    for return_node in inferred_call._get_return_nodes_skip_functions()
                 ):
                    break
```
Am I missing something ?

Reply:
Yes, I don't love this about Python.
```python
>>> all([])
True
```

Reply:
oh hang on... let me see what you did...

Reply:
I guess I need to see a suggestion in a fuller context to see if it would work. How does the suggestion track the case were the generator is empty?

Reply:
(I'm learning of this at the same time as you but) I guess the default value for an empty list can be defined by using any or all (all will be true whatever the condition, any will be false whatever the condition):
```
>>> all(i == 0 for i in range(0))
True
>>> all(i != 0 for i in range(0))
True
>>> any(i == 0 for i in range(0))
False
>>> any(i != 0 for i in range(0))
False
```

Reply:
I'd love for this to be true, but taking a step back:
1. first try: I want to use `all()`
2. i can't unless I use two iterators or use boolean flags, because `all` is True when empty, and I need False
3. so I rewrite `all` as `not any(not x)`
4. but I'm back to the same problem, since I've negated the expression, I need True for empty containers now
5. but `not any` gives me False for empty containers 🤕 
6. Plus "any" requires a complete scan, so we lose the short-circuiting 📉

Reply:
Hmm, let me try that offline and report back 😄 

> Plus "any" requires a complete scan, so we lose the short-circuiting

In ``all(isinstance(v, nodes.GeneratorExp) for v in x)`` we shortcut the first time ``isinstance(v, nodes.GeneratorExp)`` is False because all values needs to be true. In ``any(not isinstance(v, nodes.GeneratorExp) for v in x)`` we shortcut the first time ``not isinstance(v, nodes.GeneratorExp) `` is True because any single true value will makes any True. So I think we shortcut for the same reason i.e. ``isinstance(v, nodes.GeneratorExp)`` is False or ``not isinstance(v, nodes.GeneratorExp)`` is True which are the same condition.

Reply:
Ha of course you're right ``not any gives me False for empty containers``. I guess the python convention force us to handle the empty container ourselves in this case and using a for loop with break is the best we can do in term of performance.

Reply:
Ah yes good point about the short circuiting!